File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    static final int MAX = 100_000;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());

        if (N == K) {
            System.out.print(0);
            return;
        }

        System.out.print(bfs(N, K));
    }

    private static int bfs(int n, int k) {
        boolean[] visited = new boolean[MAX + 1];
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{n, 0});
        visited[n] = true;

        while (!queue.isEmpty()) {
            int[] current = queue.poll();

            int[] nextPos = {current[0] - 1, current[0] + 1, current[0] * 2};

            for (int next : nextPos) {
                if (next < 0 || next > MAX || visited[next]) {
                    continue;
                }

                if (next == k) {
                    return current[1] + 1;
                }

                queue.offer(new int[]{next, current[1] + 1});
                visited[next] = true;
            }
        }
        return -1;
    }
}

File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

lateinit var dist: IntArray

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val (n, k) = readln().split(" ").map { it.toInt() }
    dist = IntArray(100001) { -1 }
    dist[n] = 0
    val que: Queue<Int> = LinkedList()
    que.add(n)
    while (dist[k] == -1) {
        val cur = que.first()
        que.poll()
        if (cur - 1 >= 0 && dist[cur - 1] == -1) {
            que.add(cur - 1)
            dist[cur - 1] = dist[cur] + 1
        }
        if (cur + 1 < dist.size && dist[cur + 1] == -1) {
            que.add(cur + 1)
            dist[cur + 1] = dist[cur] + 1
        }
        if (cur * 2 < dist.size && dist[cur * 2] == -1) {
            que.add(cur * 2)
            dist[cur * 2] = dist[cur] + 1
        }
    }
    println(dist[k])
}